File: skills/root-analysis/scripts/r_templates/plotting.R
create_color_maps <- function(sample_names, color_scheme = "high_contrast", ordered_samples = NULL) {
  # 确保sample_names为字符类型，避免因子处理问题
  if (is.factor(sample_names)) {
    sample_names <- as.character(sample_names)
  }
  
  # 定义扩展的颜色方案，包含更多视觉上可区分的颜色
  color_schemes <- list(
    default = c(
      # 原有的12种颜色
      "#808080", "#8fb79d", "#dd3125", "#92bee1", 
      "#4c74b1", "#fc8c5a", "#6BAED6", "#969696",
      "#7f432a", "#a6cee3", "#1f78b4", "#b2df8a",
      # 新增的20种颜色，确保视觉区分度
      "#e31a1c", "#ff7f00", "#ff6b6b", "#33a02c",
      "#6a3d9a", "#b15928", "#4ecdc4", "#fdbf6f",
      "#45b7d1", "#cab2d6", "#ffff99", "#fb9a99",
      "#d62728", "#ff9800", "#2196f3", "#4caf50",
      "#9c27b0", "#795548", "#607d8b", "#e91e63",
      "#009688", "#cddc39", "#ffc107", "#673ab7",
      "#3f51b5", "#00bcd4", "#8bc34a", "#ff5722",
      "#9e9e9e", "#f44336", "#96ceb4", "#feca57"
    ),
    
    blue = c("#deebf7", "#c6dbef", "#9ecae1", "#6baed6",
             "#4292c6", "#2171b5", "#08519c", "#08306b",
             "#1e3a8a", "#1e40af", "#2563eb", "#3b82f6",
             "#60a5fa", "#93c5fd", "#bfdbfe", "#dbeafe"),
    
    green = c("#edf8e9", "#c7e9c0", "#a1d99b", "#74c476",
              "#41ab5d", "#238b45", "#006d2c", "#00441b",
              "#064e3b", "#065f46", "#047857", "#059669",
              "#10b981", "#34d399", "#6ee7b7", "#a7f3d0"),
    
    qualitative = c("#e41a1c", "#377eb8", "#4daf4a", "#984ea3",
                    "#ff7f00", "#ffff33", "#a65628", "#f781bf",
                    "#999999", "#66c2a5", "#fc8d62", "#8da0cb",
                    "#e78ac3", "#a6d854", "#ffd92f", "#e5c494"),
    
    # 高视觉区分度色彩方案 - 专为样本区分设计
    high_contrast = c("#E31A1C", "#1F78B4", "#33A02C", "#6A3D9A", 
                      "#FF7F00", "#dede8b", "#A65628", "#F781BF",
                      "#00CED1", "#006400", "#4B0082", "#FF4500",
                      "#DC143C", "#4169E1", "#228B22", "#8A2BE2",
                      "#FF8C00", "#FFD700", "#8B4513", "#FF1493",
                      "#0080ff", "#32CD32", "#9400D3", "#FF6347")
  )
  
  # 选择颜色方案 - 默认使用高区分度方案
  column_colors <- color_schemes[[color_scheme]] %||% color_schemes[["high_contrast"]]
  
  if (is.null(ordered_samples)) {
    col0_samples <- sample_names[sample_names == "Col_0"]
    oe_samples <- grep("OE", sample_names, value = TRUE)
    other_samples <- setdiff(sample_names, c(col0_samples, oe_samples))
    ordered_samples <- c(col0_samples, sort(other_samples), sort(oe_samples))
  } else {
    if (is.factor(ordered_samples)) {
      ordered_samples <- as.character(ordered_samples)
    }
    ordered_samples <- unique(ordered_samples)
    ordered_samples <- ordered_samples[ordered_samples %in% sample_names]
    missing <- setdiff(sample_names, ordered_samples)
    if (length(missing) > 0) {
      ordered_samples <- c(ordered_samples, missing)
    }
  }
  
  # 确保颜色数量足够 - 如果样本数量仍然超过颜色数量，生成额外的颜色
  if (length(ordered_samples) > length(column_colors)) {
    # 使用HSV颜色空间生成额外的颜色
    additional_colors_needed <- length(ordered_samples) - length(column_colors)
    
    # 生成均匀分布的色相值
    hue_values <- seq(0, 1, length.out = additional_colors_needed + 1)[1:additional_colors_needed]
    
    # 生成额外的颜色，使用不同的饱和度和明度确保区分度
    additional_colors <- hsv(h = hue_values, 
                           s = rep(c(0.8, 0.6, 0.9), length.out = additional_colors_needed),
                           v = rep(c(0.8, 0.9, 0.7), length.out = additional_colors_needed))
    
    # 合并颜色
    column_colors <- c(column_colors, additional_colors)
    
    cat("已为", length(ordered_samples), "个样本生成", length(column_colors), "种独特颜色\n")
  } else {
    cat("使用预定义颜色为", length(ordered_samples), "个样本分配颜色\n")
  }
  
  color_maps <- list(
    column = setNames(column_colors[1:length(ordered_samples)], ordered_samples),
    point = setNames(column_colors[1:length(ordered_samples)], ordered_samples)
  )
  
  # 强制确保Col_0获得灰色 (#808080) - 修复颜色分配问题
  if ("Col_0" %in% names(color_maps$column)) {
    cat("=== Col_0颜色强制修复 ===\n")
    cat("修复前Col_0颜色:", color_maps$column["Col_0"], "\n")
    
    # 强制设置Col_0为灰色
    color_maps$column["Col_0"] <- "#808080"
    color_maps$point["Col_0"] <- "#808080"
    
    cat("修复后Col_0颜色:", color_maps$column["Col_0"], "\n")
    cat("=== 修复完成 ===\n")
  } else {
    cat("警告：在样本中未找到Col_0\n")
  }
  
  # 打印详细调试信息
  cat("\n=== 详细调试信息 ===\n")
  cat("样本总数:", length(ordered_samples), "\n")
  cat("颜色总数:", length(column_colors), "\n")
  print("样本顺序：")
  print(ordered_samples)
  print("颜色映射（column）：")
  print(color_maps$column)
  
  # 特别检查Col_0的颜色分配
  if ("Col_0" %in% names(color_maps$column)) {
    cat("\n=== Col_0颜色验证 ===\n")
    cat("Col_0分配的颜色:", color_maps$column["Col_0"], "\n")
    cat("是否为灰色(#808080):", color_maps$column["Col_0"] == "#808080", "\n")
  }
  
  return(color_maps)
}

create_root_length_plot <- function(data, summary_stats, color_maps) {
  # 动态排序treatment：Mock优先（如果存在），其他按字母顺序
  all_treatments <- unique(c(as.character(data$treatment), as.character(summary_stats$treatment)))
  if ("Mock" %in% all_treatments) {
    treatment_order <- c("Mock", sort(setdiff(all_treatments, "Mock")))
  } else {
    treatment_order <- sort(all_treatments)
  }
  data$treatment <- factor(data$treatment, levels = treatment_order)
  summary_stats$treatment <- factor(summary_stats$treatment, levels = treatment_order)
  
  # 确保样本顺序与颜色映射一致
  data$sample <- factor(data$sample, levels = names(color_maps$column))
  summary_stats$sample <- factor(summary_stats$sample, levels = names(color_maps$column))
  
  # 计算每个treatment和sample组合的最大值位置
  sample_max_positions <- data %>%
    group_by(treatment, sample) %>%
    summarise(
      max_value = max(length),
      .groups = "drop"
    )
  
  # 合并最大值位置到summary_stats
  summary_stats <- summary_stats %>%
    left_join(sample_max_positions, by = c("treatment", "sample")) %>%
    mutate(
      letter_y_pos = pmax(max_value, mean_length + se) * 1.1  # 使用每个分组的最大值
    )
  
  # 计算整体y轴范围
  y_max <- max(summary_stats$letter_y_pos) * 1.1
  
  # 计算x轴范围
  n_samples <- length(unique(data$sample))
  x_min <- 0.5
  x_max <- n_samples + 0.5
  
  ggplot() +
    # 添加均值柱状图
    geom_col(data = summary_stats, 
             aes(x = sample, y = mean_length, fill = sample, color = sample),
             linewidth = 0.8,  # 修改: 使用 linewidth 替代 size
             alpha = 0.5, 
             width = 0.7) +
    
    # 添加原始数据散点
    geom_point(data = data,
               aes(x = sample, y = length, color = sample),
               position = position_jitter(width = 0.2),
               alpha = 0.9, 
               size = 1) +
    
    # 添加误差线和其他图形元素
    geom_errorbar(data = summary_stats,
                  aes(x = sample,
                      ymin = mean_length - se,
                      ymax = mean_length + se),
                  width = 0.2) +
    
    # 添加显著性字母标记
    geom_text(data = summary_stats,
              aes(x = sample,
                  y = letter_y_pos),  # 使用每个样本的独立位置
              label = summary_stats$letter,
              size = 6,
              vjust = -0.5) +
    
    # 设置主题和样式
    theme_classic() +
    labs(x = "Sample", y = "Root length (cm)") +
    apply_plot_theme() +
    
    # 设置颜色 - 使用传入的颜色映射
    scale_fill_manual(values = color_maps$column) +
    scale_color_manual(values = color_maps$point) +
    
    # 分面
    facet_wrap(~ treatment) +
    
    # 控制Y轴范围和扩展
    scale_y_continuous(
      limits = c(0, y_max),
      expand = expansion(mult = c(0, 0))
    ) +
    
    # 控制X轴扩展
    scale_x_discrete(
      expand = expansion(mult = c(0.2, 0.05))
    ) +
    
    # 修改坐标系
    coord_cartesian(
      xlim = c(x_min, x_max),
      clip = "off",
      expand = FALSE
    ) +
    
    # 添加主题调整确保轴线完整显示
    theme(
      axis.line = element_line(linewidth = 0.5, colour = "black"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank()
    )
}

apply_plot_theme <- function() {
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
    axis.text.y = element_text(size = 12),
    axis.title = element_text(size = 14),
    plot.title = element_text(hjust = 0.5, vjust = 1),
    legend.position = "none",
    legend.text = element_text(size = 12, face = "plain"),
    legend.title = element_text(size = 14, face = "bold"),
    strip.text.x = element_text(size = 14)
  )
}


# 创建比率图
create_ratio_plot <- function(ratios_data, ratio_letters) {
  # 准备数据
  ratio_summary <- ratios_data %>%
    filter(treatment != "Mock") %>%  # 移除Mock组
    group_by(sample) %>%
    summarise(
      mean_ratio = mean(ratio, na.rm = TRUE),
      se = sd(ratio, na.rm = TRUE) / sqrt(n()),
      .groups = "drop"
    ) %>%
    left_join(ratio_letters, by = "sample")
  
  # 设置颜色
  sample_colors <- c("#ffde91", "#8fb79d", "#dd3125")
  names(sample_colors) <- unique(ratios_data$sample)
  
  # 创建图形
  y_max <- max(ratio_summary$mean_ratio + ratio_summary$se) + 0.2
  
  plot <- ggplot() +
    # 添加柱状图
    geom_col(data = ratio_summary,
             aes(x = sample, y = mean_ratio, fill = sample),
             alpha = 0.5,
             width = 0.7) +
    
    # 添加误差线
    geom_errorbar(data = ratio_summary,
                  aes(x = sample,
                      ymin = mean_ratio - se,
                      ymax = mean_ratio + se),
                  width = 0.2) +
    
    # 添加显著性字母
    geom_text(data = ratio_summary,
              aes(x = sample,
                  y = mean_ratio + se + 0.1,
                  label = letter),
              size = 6) +
    
    # 设置主题和标签
    theme_classic() +
    labs(x = "Sample",
         y = "Relative root length (treatment/mock)") +
    
    # 设置颜色
    scale_fill_manual(values = sample_colors) +
    
    # 控制坐标轴
    scale_y_continuous(
      limits = c(0, y_max),
      expand = c(0, 0)
    ) +
    
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
      axis.text.y = element_text(size = 12),
      axis.title = element_text(size = 14),
      legend.position = "none"
    )
  
  return(plot)
}

create_ratio_plot_with_points <- function(ratios_data, letters_df, color_maps) {
  # 打印调试信息
  print("原始样本顺序：")
  print(unique(ratios_data$sample))
  print("颜色映射中的样本顺序：")
  print(names(color_maps$column))
  
  # 确保样本顺序与颜色映射一致
  ratios_data$sample <- factor(ratios_data$sample, levels = names(color_maps$column))
  
  # 打印转换后的顺序
  print("转换为因子后的样本顺序：")
  print(levels(ratios_data$sample))
  
  # 计算统计摘要
  summary_stats <- ratios_data %>%
    group_by(sample) %>%
    summarise(
      mean_ratio = mean(ratio, na.rm = TRUE),
      se = sd(ratio, na.rm = TRUE) / sqrt(n()),
      max_value = max(ratio),
      .groups = "drop"
    ) %>%
    left_join(letters_df, by = "sample") %>%
    mutate(
      letter_y_pos = pmax(max_value, mean_ratio + se) * 1.1,
      # 重要：确保保持正确的顺序
      sample = factor(sample, levels = names(color_maps$column))
    )
  
  # 打印最终的统计数据顺序
  print("统计数据中的样本顺序：")
  print(levels(summary_stats$sample))
  
  # 计算整体y轴范围
  y_max <- max(summary_stats$letter_y_pos) * 1.1
  
  # 计算x轴范围
  n_samples <- length(unique(ratios_data$sample))
  x_min <- 0.5
  x_max <- n_samples + 0.5
  
  ggplot() +
    # 添加柱状图
    geom_col(data = summary_stats,
             aes(x = sample, y = mean_ratio, fill = sample, color = sample),
             linewidth = 0.8,
             alpha = 0.5,
             width = 0.7) +
    
    # 添加散点
    geom_point(data = ratios_data,
               aes(x = sample, y = ratio, color = sample),
               position = position_jitter(width = 0.2),
               alpha = 0.9,
               size = 2) +
    
    # 添加误差线
    geom_errorbar(data = summary_stats,
                  aes(x = sample,
                      ymin = mean_ratio - se,
                      ymax = mean_ratio + se),
                  width = 0.2) +
    
    # 添加显著性字母
    geom_text(data = summary_stats,
              aes(x = sample,
                  y = letter_y_pos),  # 使用每个样本的独立位置
              label = summary_stats$letter,
              size = 6,
              vjust = -0.5) +
    
    # 设置主题和标签
    theme_classic() +
    labs(x = NULL,
         y = "Root length ratio (ISX/Mock)") +
    apply_plot_theme() +
    
    # 设置颜色 - 使用传入的颜色映射
    scale_fill_manual(values = color_maps$column) +
    scale_color_manual(values = color_maps$point) +
    
    # 设置y轴范围
    scale_y_continuous(
      limits = c(0, y_max),
      expand = expansion(mult = c(0, 0))
    ) +
    
    # 控制X轴扩展
    scale_x_discrete(
      expand = expansion(mult = c(0.2, 0.05))
    ) +
    
    # 修改坐标系
    coord_cartesian(
      xlim = c(x_min, x_max),
      clip = "off",
      expand = FALSE
    ) +
    
    # 添加主题调整
    theme(
      axis.line = element_line(linewidth = 0.5, colour = "black"),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_blank(),
      panel.background = element_blank(),
      axis.title.x = element_blank(), # 隐藏 X 轴标题
      #axis.text.x = element_blank()   # 隐藏 X 轴刻度标签
    )
}

File: .nini/skills/root-analysis/scripts/python_templates/plotting.py
"""
绘图模块 - Python版本

使用matplotlib和seaborn生成出版级图表
"""

import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import seaborn as sns
import pandas as pd
import numpy as np
from pathlib import Path


# 设置中文字体支持
plt.rcParams['font.sans-serif'] = ['SimHei', 'DejaVu Sans', 'Arial Unicode MS', 'Arial']
plt.rcParams['axes.unicode_minus'] = False  # 解决负号显示问题


class RootLengthPlotter:
    """根长度数据绘图器"""

    # 颜色方案定义
    COLOR_SCHEMES = {
        'high_contrast': [
            '#E31A1C', '#1F78B4', '#33A02C', '#6A3D9A',
            '#FF7F00', '#DEDE8B', '#A65628', '#F781BF',
            '#00CED1', '#006400', '#4B0082', '#FF4500',
            '#DC143C', '#4169E1', '#228B22', '#8A2BE2',
            '#FF8C00', '#FFD700', '#8B4513', '#FF1493',
            '#0080FF', '#32CD32', '#9400D3', '#FF6347'
        ],
        'default': [
            '#808080', '#8FB79D', '#DD3125', '#92BEE1',
            '#4C74B1', '#FC8C5A', '#6BAED6', '#969696',
            '#7F432A', '#A6CEE3', '#1F78B4', '#B2DF8A',
            '#E31A1C', '#FF7F00', '#FF6B6B', '#33A02C',
            '#6A3D9A', '#B15928', '#4ECDC4', '#FDBF6F'
        ],
        'blue': [
            '#DEEBF7', '#C6DBEF', '#9ECAE1', '#6BAED6',
            '#4292C6', '#2171B5', '#08519C', '#08306B',
            '#1E3A8A', '#1E40AF', '#2563EB', '#3B82F6',
            '#60A5FA', '#93C5FD', '#BFDBFE', '#DBEAFE'
        ],
        'green': [
            '#EDF8E9', '#C7E9C0', '#A1D99B', '#74C476',
            '#41AB5D', '#238B45', '#006D2C', '#00441B',
            '#064E3B', '#065F46', '#047857', '#059669',
            '#10B981', '#34D399', '#6EE7B7', '#A7F3D0'
        ],
        'qualitative': [
            '#E41A1C', '#377EB8', '#4DAF4A', '#984EA3',
            '#FF7F00', '#FFFF33', '#A65628', '#F781BF',
            '#999999', '#66C2A5', '#FC8D62', '#8DA0CB',
            '#E78AC3', '#A6D854', '#FFD92F', '#E5C494'
        ]
    }

    def __init__(self, color_scheme='high_contrast'):
        """
        初始化绘图器

        Args:
            color_scheme: 颜色方案名称
        """
        self.color_scheme = color_scheme
        self.colors = self.COLOR_SCHEMES.get(color_scheme, self.COLOR_SCHEMES['high_contrast'])

    def create_color_map(self, samples, ordered_samples=None):
        """
        创建样本到颜色的映射

        Args:
            samples: 样本列表
            ordered_samples: 可选的样本排序

        Returns:
            dict: {sample: color}
        """
        # 排序样本
        if ordered_samples is not None:
            sample_order = [s for s in ordered_samples if s in samples]
            missing = [s for s in samples if s not in ordered_samples]
            sample_order.extend(sorted(missing))
        else:
            # 自动排序：Col_0优先，非OE样本，OE样本
            col0 = [s for s in samples if s == 'Col_0']
            oe_samples = [s for s in samples if 'OE' in str(s) and s != 'Col_0']
            other_samples = [s for s in samples if s not in col0 and s not in oe_samples]
            sample_order = col0 + sorted(other_samples) + sorted(oe_samples)

        # 确保有足够的颜色
        if len(sample_order) > len(self.colors):
            # 生成额外颜色
            extra_colors = self._generate_extra_colors(len(sample_order) - len(self.colors))
            colors = self.colors + extra_colors
        else:
            colors = self.colors

        # 创建映射
        color_map = {sample: colors[i] for i, sample in enumerate(sample_order)}

        # Col_0强制使用灰色
        if 'Col_0' in color_map:
            color_map['Col_0'] = '#808080'

        return color_map, sample_order

    def _generate_extra_colors(self, n):
        """生成额外的颜色"""
        colors = []
        for i in range(n):
            hue = i / n
            rgb = plt.cm.hsv(hue)
            hex_color = '#{:02x}{:02x}{:02x}'.format(
                int(rgb[0] * 255),
                int(rgb[1] * 255),
                int(rgb[2] * 255)
            )
            colors.append(hex_color)
        return colors

    def plot_root_length(self, data, summary_stats, color_map, sample_order,
                        output_path, figsize=(10, 6)):
        """
        绘制根长度图（分面柱状图+散点+误差线+显著性字母）

        Args:
            data: 原始数据DataFrame
            summary_stats: 统计摘要DataFrame
            color_map: 颜色映射字典
            sample_order: 样本顺序列表
            output_path: 输出文件路径
            figsize: 图表尺寸
        """
        # 获取处理组
        treatments = sorted(data['treatment'].unique(),
                          key=lambda x: (x != 'Mock', x))  # Mock优先

        n_treatments = len(treatments)

        # 创建子图
        fig, axes = plt.subplots(1, n_treatments, figsize=figsize,
                                sharey=True, squeeze=False)
        axes = axes.flatten()

        for idx, treatment in enumerate(treatments):
            ax = axes[idx]

            # 筛选数据
            treatment_data = data[data['treatment'] == treatment]
            treatment_summary = summary_stats[summary_stats['treatment'] == treatment]

            # 确保样本顺序
            treatment_summary = treatment_summary.set_index('sample').loc[
                [s for s in sample_order if s in treatment_summary['sample'].values]
            ].reset_index()

            x_pos = np.arange(len(treatment_summary))

            # 绘制柱状图
            bars = ax.bar(x_pos, treatment_summary['mean'],
                         color=[color_map[s] for s in treatment_summary['sample']],
                         alpha=0.5, edgecolor='black', linewidth=1.2,
                         label='Mean')

            # 绘制散点
            for i, sample in enumerate(treatment_summary['sample']):
                sample_data = treatment_data[treatment_data['sample'] == sample]
                y_values = sample_data['length'].values
                x_values = np.random.normal(i, 0.04, len(y_values))  # 抖动
                ax.scatter(x_values, y_values,
                          color=color_map[sample], alpha=0.8,
                          s=30, zorder=3, edgecolors='black', linewidth=0.5)

            # 绘制误差线
            ax.errorbar(x_pos, treatment_summary['mean'],
                       yerr=treatment_summary['sem'],
                       fmt='none', ecolor='black', capsize=4,
                       linewidth=1.5, zorder=4)

            # 添加显著性字母
            y_max = (treatment_summary['mean'] + treatment_summary['sem']).max()
            y_pos = y_max * 1.15

            for i, (_, row) in enumerate(treatment_summary.iterrows()):
                if pd.notna(row['letter']) and row['letter']:
                    ax.text(i, y_pos, row['letter'],
                           ha='center', va='bottom',
                           fontsize=14, fontweight='bold')

            # 设置标签和标题
            ax.set_xlabel('Sample', fontsize=12, fontweight='bold')
            if idx == 0:
                ax.set_ylabel('Root length (cm)', fontsize=12, fontweight='bold')
            ax.set_title(treatment, fontsize=14, fontweight='bold')

            # 设置x轴
            ax.set_xticks(x_pos)
            ax.set_xticklabels(treatment_summary['sample'],
                              rotation=45, ha='right')

            # 网格
            ax.grid(axis='y', alpha=0.3, linestyle='--')
            ax.set_axisbelow(True)

            # 美化
            ax.spines['top'].set_visible(False)
            ax.spines['right'].set_visible(False)

        # 调整布局
        plt.tight_layout()

        # 保存
        output_path = Path(output_path)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        plt.savefig(output_path, dpi=300, bbox_inches='tight')
        plt.close()

        print(f"根长度图已保存: {output_path}")

    def plot_ratio(self, ratio_df, ratio_letters, color_map, sample_order,
                  output_path, figsize=(8, 6)):
        """
        绘制比率图

        Args:
            ratio_df: 比率数据DataFrame
            ratio_letters: 显著性字母字典
            color_map: 颜色映射
            sample_order: 样本顺序
            output_path: 输出路径
            figsize: 图表尺寸
        """
        if len(ratio_df) == 0:
            print("跳过比率图（无数据）")
            return

        # 计算统计量
        summary = ratio_df.groupby('sample').agg({
            'ratio': ['mean', 'sem', 'max']
        }).reset_index()
        summary.columns = ['sample', 'mean', 'sem', 'max']

        # 添加字母
        summary['letter'] = summary['sample'].map(ratio_letters)

        # 排序
        summary = summary.set_index('sample').loc[
            [s for s in sample_order if s in summary.index]
        ].reset_index()

        # 创建图表
        fig, ax = plt.subplots(figsize=figsize)

        x_pos = np.arange(len(summary))

        # 柱状图
        bars = ax.bar(x_pos, summary['mean'],
                     color=[color_map[s] for s in summary['sample']],
                     alpha=0.5, edgecolor='black', linewidth=1.2)

        # 散点
        for i, sample in enumerate(summary['sample']):
            sample_data = ratio_df[ratio_df['sample'] == sample]
            y_values = sample_data['ratio'].values
            x_values = np.random.normal(i, 0.04, len(y_values))
            ax.scatter(x_values, y_values,
                      color=color_map[sample], alpha=0.8,
                      s=40, zorder=3, edgecolors='black', linewidth=0.5)

        # 误差线
        ax.errorbar(x_pos, summary['mean'], yerr=summary['sem'],
                   fmt='none', ecolor='black', capsize=4,
                   linewidth=1.5, zorder=4)

        # 显著性字母
        y_max = (summary['mean'] + summary['sem']).max()
        y_pos = y_max * 1.15

        for i, (_, row) in enumerate(summary.iterrows()):
            if pd.notna(row['letter']) and row['letter']:
                ax.text(i, y_pos, row['letter'],
                       ha='center', va='bottom',
                       fontsize=14, fontweight='bold')

        # 参考线（ratio=1）
        ax.axhline(y=1.0, color='red', linestyle='--',
                  linewidth=1.5, alpha=0.7, label='Baseline (ratio=1)')

        # 标签
        ax.set_xlabel('Sample', fontsize=12, fontweight='bold')
        ax.set_ylabel('Root length ratio (Treatment/Mock)', fontsize=12, fontweight='bold')
        ax.set_title('Relative Root Length Analysis', fontsize=14, fontweight='bold')

        # x轴
        ax.set_xticks(x_pos)
        ax.set_xticklabels(summary['sample'], rotation=45, ha='right')

        # 网格
        ax.grid(axis='y', alpha=0.3, linestyle='--')
        ax.set_axisbelow(True)

        # 美化
        ax.spines['top'].set_visible(False)
        ax.spines['right'].set_visible(False)
        ax.legend(loc='best', frameon=False)

        # 保存
        plt.tight_layout()
        output_path = Path(output_path)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        plt.savefig(output_path, dpi=300, bbox_inches='tight')
        plt.close()

        print(f"比率图已保存: {output_path}")

    def save_color_mapping(self, color_map, sample_order, output_path):
        """保存颜色映射到CSV"""
        df = pd.DataFrame({
            'sample': sample_order,
            'color': [color_map[s] for s in sample_order]
        })

        output_path = Path(output_path)
        output_path.parent.mkdir(parents=True, exist_ok=True)
        df.to_csv(output_path, index=False)

        print(f"颜色映射已保存: {output_path}")

    def save_sample_order(self, sample_order, output_path):
        """保存样本顺序到文本文件"""
        lines = ["样本排列顺序（左→右）:"]
        for i, sample in enumerate(sample_order, 1):
            lines.append(f"{i}. {sample}")

        output_path = Path(output_path)
        output_path.parent.mkdir(parents=True, exist_ok=True)

        with open(output_path, 'w', encoding='utf-8') as f:
            f.write('\n'.join(lines))

        print(f"样本顺序已保存: {output_path}")
